src/components/annual-report/section3/useFetchSectionThreeData.js

```diff
@@ -0,0 +1,126 @@
+
+            const externalRelations =
+                variables?.mappedConfiguration?.externalRelations
+            const externalRelationsDEs = externalRelations
+                ? [
+                      ...externalRelations.map((rel) => rel.denominator),
+                      ...externalRelations.map((rel) => rel.numerator),
+                      ...externalRelations.map((rel) => rel.externalData),
+                  ]
+                : []
+
+            // set to loading
+            setLoading(true)
+            // get data sets and check that each data element is associated with only one period type
+
+            // we need to preserve information on index to tie response back to relevant denominator relation (some can be filtered out)
+            const byLevelRequestsWithIndices = externalRelations
+                .map((rel, index) => {
+                    // if the relation's level is the same, or less the boundary organisation unit level, the byLevel request should be skipped
+                    if (!rel.level || rel.level <= boundaryOrgUnitLevel) {
+                        return null
+                    }
+
+                    return {
+                        request: fetchDataByLevel({
+                            engine,
+                            variables: {
+                                dataElements: [
+                                    rel.numerator,
+                                    rel.denominator,
+                                    rel.externalData,
+                                ],
+                                orgUnits: variables.orgUnits,
+                                orgUnitLevel: `LEVEL-${rel.level}`,
+                                periods: [variables?.currentPeriod.id],
+                            },
+                        }),
+                        index,
+                    }
+                })
+                .filter((req) => req) //filter out null requests
+            const byLevelRequests = byLevelRequestsWithIndices.map(
+                (r) => r.request
+            )
+            const byLevelRequestIndices = byLevelRequestsWithIndices.map(
+                (r) => r.index
+            )
+            try {
+                const overallRequest = engine.query(section3Overall, {
+                    variables: {
+                        dataElements: externalRelationsDEs,
+                        orgUnits: variables.orgUnits,
+                        periods: [variables.currentPeriod.id],
+                    },
+                })
+                const [overallData, ...dataByLevel] = await Promise.all([
+                    overallRequest,
+                    ...byLevelRequests,
+                ])
+
+                setData({
+                    ...overallData,
+                    data_detail_by_level: dataByLevel.map(
+                        (resp) => resp?.data_detail_by_level
+                    ),
```

Comment:
(also added a suggestion in section 4)

```suggestion
                    [BY_LEVEL_RESPONSE_NAME]: dataByLevel.map(
                        (resp) => resp?.[BY_LEVEL_RESPONSE_NAME]
                    ),
```